### PR TITLE
EOS-15037: Object listing with prefix specified continues key enumeration even after prefix match stops

### PR DESCRIPTION
**Pull request by DPG17**
- Code changes done to stop further object listing from motr/clovis once the prefix does not
  match after it's previous match.

Signed-off-by: ‘Dattaprasad <dattaprasad.govekar@seagate.com>

### CODE REVIEW

**Comment by DPG17** [comment]
Pre merge build: http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/145/ triggered (in progress)

File: server/s3_get_bucket_action.cc [hunk 1]
```diff
@@ -214,10 +216,23 @@ void S3GetBucketAction::get_next_objects_successful() {
...
+          // further enumeration.
+          if (last_key_prefix_match) {
+            // No further prefix match will occur (as items in Motr storage are
+            // arranaged in lexical order)
+            skip_no_further_prefix_match = true;
+            // Set length to zero to indicate truncation is false
+            length = 0;
+            s3_log(
+                S3_LOG_DEBUG, request_id,
+                "No further prefix match. Skipping further object listing\n");
```
Comment (knrajnambiar76) [inline server/s3_get_bucket_action.cc:229]: Have request_prefix printed in log

**Review by knrajnambiar76 — APPROVED** [review]

File: server/s3_get_bucket_action.cc [hunk 2]
```diff
@@ -272,6 +288,20 @@ void S3GetBucketAction::get_next_objects_successful() {
...
+      } else {
+        // Prefix does not match
+        // If previous/last key was matching the prefix, stop and break further
+        // enumeration.
+        if (last_key_prefix_match) {
+          // No further prefix match will occur (as items in Motr storage are
+          // arranaged in lexical order)
+          skip_no_further_prefix_match = true;
+          // Set length to zero to indicate truncation is false
+          length = 0;
```
Comment (knrajnambiar76) [inline server/s3_get_bucket_action.cc:300]: length is set to 0, 
below  --length == 0 wont be satisfied

File: server/s3_get_bucket_action.cc (hunk 1 above)
Comment (DPG17) [reply server/s3_get_bucket_action.cc:229]: yes. it is in log at debug level

Comment (DPG17) [reply server/s3_get_bucket_action.cc:229]: do you want to print request prefix at line #227?

File: server/s3_get_bucket_action.cc [hunk 3]
```diff
@@ -214,10 +216,23 @@ void S3GetBucketAction::get_next_objects_successful() {
         // Filter out by prefix
         if (kv.first.find(request_prefix) == std::string::npos) {
           // Key does not start with specified prefix; key filetered out.
-          // Check the next key.
+          // If previous/last key was matching the prefix, stop and break
+          // further enumeration.
+          if (last_key_prefix_match) {
```
Comment (knrajnambiar76) [inline server/s3_get_bucket_action.cc:221]: you can replace last_key_prefix_match with 
kv.first > request_prefix everywhere

File: server/s3_get_bucket_action.cc (hunk 1 above)
Comment (knrajnambiar76) [reply server/s3_get_bucket_action.cc:228]: Its ok

**Review by knrajnambiar76 — APPROVED** [review]
Later please have a look to how to optimize system tests, as its taking around 15 minutes

**Comment by nileshgovande** [comment]
Pre-merge job - http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/153/

**Review by amitwac2608 — APPROVED** [review]

**Comment by DPG17** [comment]
Pre merge build: http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/155/ triggered (in progress)

**Comment by DPG17** [comment]
Triggered pre merge job: http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/159/ (in progress)

**Review by nileshgovande — APPROVED** [review]

File: server/s3_get_bucket_action.cc [hunk 4]
```diff
@@ -272,6 +286,19 @@ void S3GetBucketAction::get_next_objects_successful() {
...
+      } else {
+        // Prefix does not match.
+        // Check if fetched key is lexicographically greater than prefix
+        if (kv.first > request_prefix) {
+          // No further prefix match will occur (as items in Motr storage are
+          // arranaged in lexical order)
+          skip_no_further_prefix_match = true;
+          // Set length to zero to indicate truncation is false
+          length = 0;
+          s3_log(S3_LOG_INFO, request_id,
```
[the remark this answers is not in the document]
Comment (DPG17) [reply server/s3_get_bucket_action.cc:309]: This log entry will not flood S3 server log. For one list object request, there is only one such entry.

File: server/s3_get_bucket_action.cc (hunk 3 above)
Comment (DPG17) [reply server/s3_get_bucket_action.cc]: New changes compare string lexicographically

**Comment by DPG17** [comment]
For this change, no changes are done in ST.  Accidently, I used 'st/clitests/awss3api_spec.py' from another branch/commit with changes to test object listing with hierarchical keys, as these changes were not available for 'br/DG/EOS-15037' local branch. Later, it seems the file got committed to local branch (accidently) and merge happened from upstream branch 'origin/br/DG/EOS-15037' to my local branch, which created duplicate code in file 'st/clitests/awss3api_spec.py'
I've now reverted the previously pushed content of 'st/clitests/awss3api_spec.py' to same content as main. So, net-net, there is no change in ST (st/clitests/awss3api_spec.py) for this review.. There may be a single blank line of change; one can ignore that change..

**Pull request merged**